### PR TITLE
SAK-43610: samigo > time exceptions > add validation to time limit

### PR DESCRIPTION
**Pull request by bjones86**
https://jira.sakaiproject.org/browse/SAK-43610

Adds validation and error messaging when time extension time limit is greater than availability window.

Took the opportunity to reduce code duplication between `ConfirmPublishAssessmentListener`, `SavePublishedSettingsListener`, `AssessmentSettingsBean`, and `PublishedAssessmentSettingsBean` by introducing a new delegate class, `ExtendedTimeValidator`.

Most of the code in this new delegate is just moved from the other classes. I'll denote inline where the new logic lives.

### CODE REVIEW

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/util/ExtendedTimeValidator.java [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+
+                long openWindowMinutes = (due.getTime() - entry.getStartDate().getTime()) / 1000 / 60;
+                if( openWindowMinutes < timerMinutes )
+                {
+                    String errorString = getError( ERROR_KEY_OPEN_WINDOW_LESS_THAN_LIMIT, entry );
+                    context.addMessage( null, new FacesMessage( FacesMessage.SEVERITY_WARN, errorString, null ) );
+                    valid = false;
+                }
+            }
+        }
```
Comment (bjones86) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/util/ExtendedTimeValidator.java:183]: This is the newly introduced validation routine.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/util/ExtendedTimeValidator.java [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+            Date startDate = entry.getStartDate();
+            Date dueDate = entry.getDueDate();
+            Date retractDate = entry.getRetractDate();
+
+            if( StringUtils.isNotEmpty( user ) )
+            {
+                users.add( user );
+            }
+
+            if( StringUtils.isNotEmpty( group ) )
```
Comment (ottenhoff) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/util/ExtendedTimeValidator.java:107]: i occassionally see instructors choose both user and group .... does it make sense to just assume that they mean one user and ignore the group selection ?

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/util/ExtendedTimeValidator.java [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+                    dupUsers = "'" + getUserName( entry ) + "'";
+                }
+                else if( count < (end - 1) )
+                {
+                    dupUsers = dupUsers + ", '" + getUserName( entry ) + "'";
+                }
+                else
+                {
+                    String and = ContextUtil.getLocalizedString( ASSESSMENT_SETTINGS_BUNDLE, MSG_KEY_AND );
+                    dupUsers = dupUsers + ", " + and + " '" + getUserName( entry );
```
Comment (ottenhoff) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/util/ExtendedTimeValidator.java:204]: wow this is a lot of lifting for duplicates!

**Review by ottenhoff — APPROVED** [review]

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/util/ExtendedTimeValidator.java (hunk 2 above)
Comment (bjones86) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/util/ExtendedTimeValidator.java:107]: This is the original logic. In my testing, you can still select either a user, or a group, or a user and a group.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/util/ExtendedTimeValidator.java (hunk 3 above)
Comment (bjones86) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/util/ExtendedTimeValidator.java:204]: Yeah I didn't write most of this code. Some could certainly be improved but I tried not to touch original logic.

**Comment by bjones86** [comment]
@ottenhoff I pushed another commit for a bug fix, do you mind taking another look?

**Review by ottenhoff — APPROVED** [review]
LGTM

**Pull request merged**